Network/TCP.hs

```diff
@@ -236,15 +236,35 @@ openTCPConnection_ uri port stashInput = do
     -- like this as it just does a once-only installation of a shutdown handler to run at program exit,
     -- rather than actually shutting down after the action
     addrinfos <- withSocketsDo $ getAddrInfo (Just $ defaultHints { addrFamily = AF_UNSPEC, addrSocketType = Stream }) (Just fixedUri) (Just . show $ port)
+
+    let
+      connectAddrInfo a = do
+        s <- socket (addrFamily a) Stream defaultProtocol
+        onException (do
+                    setSocketOption s KeepAlive 1
+                    connect s (addrAddress a)
+                    socketConnection_ fixedUri port s stashInput
+                    ) (Network.Socket.close s)
+
+      -- try multiple addresses; return Just connected socket or Nothing
+      tryAddrInfos [] = return Nothing
+      tryAddrInfos (h:t) = fmap Just (connectAddrInfo h) `catchIO_` tryAddrInfos t
```

Comment:
You should use `try` rather than running the remaining connections in an async-exception masked catch handler.